### PR TITLE
fix: stop agent pebble svc not exist error & validate credentials error

### PR DESCRIPTION
**Pull request by yanksyoon**
Applicable spec: N/A

### Overview

1. Catches pebble API error when `pebble.stop(service)` has been called for services not yet instantiated.
2. Remove `validate_credentials` for agent relation since there is a concrete 1:1 mapping for agent relation. When validate credential fails due to Jenkins error, the event is not retried and the agent does not get registered at all in the end.
Example Jenkins error (somehow 404):
```
2024-01-02T01:29:57.224Z [container-agent] Exception in thread "main" java.io.IOException: Failed to obtain http://10.86.69.157:8080/computer/jenkins-agent-k8s-1/slave-agent.jnlp?encrypt=true
2024-01-02T01:29:57.224Z [container-agent] 	at hudson.remoting.Launcher.parseJnlpArguments(Launcher.java:567)
2024-01-02T01:29:57.224Z [container-agent] 	at hudson.remoting.Launcher.run(Launcher.java:346)
2024-01-02T01:29:57.224Z [container-agent] 	at hudson.remoting.Launcher.main(Launcher.java:297)
2024-01-02T01:29:57.224Z [container-agent] Caused by: java.io.IOException: Failed to load http://10.86.69.157:8080/computer/jenkins-agent-k8s-1/slave-agent.jnlp?encrypt=true: 404 Not Found
2024-01-02T01:29:57.224Z [container-agent] 	at hudson.remoting.Launcher.parseJnlpArguments(Launcher.java:514)
2024-01-02T01:29:57.224Z [container-agent] 	... 2 more
2024-01-02T01:29:57.224Z [container-agent]
```

### Rationale

1. This fails relation-broken event and falls into errored state.
2. Let pebble retry failing agent registration.

### Juju Events Changes

None.

### Module Changes

`pebble.py`'s `stop_agent` method now handles pebble API error for service not exists case.

### Library Changes

None.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/pebble.py [hunk 1]
```diff
@@ -89,6 +89,14 @@ def stop_agent(self, container: ops.Container) -> None:
...
             container: The agent workload container.
+
+        Raises:
+            APIError: if something went wrong with pebble requesting service stop.
         """
-        container.stop(self.state.jenkins_agent_service_name)
+        try:
+            container.stop(self.state.jenkins_agent_service_name)
+        except ops.pebble.APIError as exc:
+            if f'service "{self.state.jenkins_agent_service_name}" does not exist' in exc.message:
```
Comment (weiiwang01) [inline src/pebble.py:99]: Perhaps using `ops.pebble.APIError.code` to determine the type of error would be more resilient. I assume the error code here is `404`.

Comment (yanksyoon) [reply src/pebble.py:99]: The logs I've encountered didn't quite include the code, only the error message was displayed, I'll validate that the error code exists and apply the fix, thanks for the suggestion!

Comment (yanksyoon) [reply src/pebble.py:99]: @weiiwang01 I'm not quite confident on using `exc.code` in this case, the err code returned is 400 here and i think pebble might not be returning the error codes by convention - we might be catching errors outside of intended scope.
Log added: `logger.error("APIERR: %s %s", exc.message, exc.code)`
```
APIERR: cannot stop services: service "jenkins-k8s-agent" does not exist 400
```

Comment (weiiwang01) [reply src/pebble.py:99]: Gotcha 👍

**Review by weiiwang01 — DISMISSED** [review]

**Review by arturo-seijas — APPROVED** [review]

File: src/pebble.py (hunk 1 above)
Comment (gregory-schiano) [reply src/pebble.py:99]: Isn't it more reliable to use `get_service` before calling stop, and only calling stop if the service is available ?
I find relying on the message of the exception to be a bit risky

Comment (yanksyoon) [reply src/pebble.py]: @gregory-schiano That is a great idea, fixed, thank you!

**Review by weiiwang01 — APPROVED** [review]

File: src/pebble.py [hunk 2]
```diff
@@ -89,14 +89,12 @@ def stop_agent(self, container: ops.Container) -> None:
...
         try:
-            container.stop(self.state.jenkins_agent_service_name)
-        except ops.pebble.APIError as exc:
-            if f'service "{self.state.jenkins_agent_service_name}" does not exist' in exc.message:
-                return
-            raise
+            # use get_service to check if service should be stopped rather than stopping and
+            # catching ops.pebble.APIError and parsing error message to determine type of error.
+            container.get_service(self.state.jenkins_agent_service_name)
+        except ops.ModelError:
```
Comment (arturo-seijas) [inline src/pebble.py:97]: What's this exception?

Comment (yanksyoon) [reply src/pebble.py:97]: This exception is raised when `container.get_service` is called and the service doesn't exist. It is a workaround for having to parse pebble's APIError :)

Comment (arturo-seijas) [reply src/pebble.py:97]: I see. Thx

**Review by yhaliaw — APPROVED** [review]

**Pull request merged**